src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java

```diff
@@ -503,6 +502,80 @@ public void completeTask(String taskId, AccessControlResponse accessControlRespo
         }
     }
 
+    private void checkPermissions(String taskId, AccessControlResponse accessControlResponse,
+            boolean isGranularPermissionFeatureEnabled, String userId) {
+        PermissionRequirements permissionsRequired;
+        if (isGranularPermissionFeatureEnabled) {
+            permissionsRequired = PermissionRequirementBuilder.builder()
+                .initPermissionRequirement(asList(OWN, EXECUTE), OR)
+                .joinPermissionRequirement(OR)
+                .nextPermissionRequirement(asList(COMPLETE), OR)
+                .build();
+        } else {
+            permissionsRequired = PermissionRequirementBuilder.builder()
+                .buildSingleRequirementWithOr(OWN, EXECUTE);
+        }
+
+        TaskResource taskResource;
+        try {
+            taskResource = roleAssignmentVerification.verifyRoleAssignments(
+                taskId, accessControlResponse.getRoleAssignments(), permissionsRequired
+            );
+        }  catch (RoleAssignmentVerificationException e) {
+            if (isGranularPermissionFeatureEnabled) {
+                taskResource = checkCompleteOwn(taskId, accessControlResponse, userId);
```

Comment:
In addition to the above comment, if you pass COMPLETE_OWN as well, you don't need to call the verifyRoleAssignments again, isn't it?